### PR TITLE
Feature/grpc pass

### PR DESCRIPTION
**Pull request by gaurav-gogia**
gRPC API Sink Tagger

This pass aims to model gRPC function calls as API calls
```java
public String greet(String name) {
    logger.info("Will try to greet " + name + " ...");
    HelloRequest request = HelloRequest.newBuilder().setName(name).build();

    // Use a stub with the given call credentials applied to invoke the RPC.
    HelloReply response =
        blockingStub
            .withCallCredentials(callCredentials)
            .sayHello(request);

    logger.info("Greeting: " + response.getMessage());
    return response.getMessage();
  }
```
https://github.com/grpc/grpc-java/blob/844de39c2640491c05b0fa6102eb46085ba7b3bf/examples/example-jwt-auth/src/main/java/io/grpc/examples/jwtauth/AuthClient.java#L71

What works?
This pass works for standardised examples found in `grpc-java` repo. Code sample shared above.

### Sample repo where this feature was tested
https://github.com/anjeyy/grpc-sample

### Sample flows
sample2
sample1

## Flags for testing
`-ead` and `-dfsde` flags were used in `scan` command to get the flows

What needs more investigation?
1. Doesn't work for custom/non-standard implementations
2. [Capn proto](https://capnproto.org/) is another example that needs to be handled separately

### CODE REVIEW

File: src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        // Get full type name of arguments going inside gRPC sink call
+        val sinkArgTypes = sinkCall.argument.isIdentifier.typeFullName.l
+        if (endpointArgTypes.toSet == sinkArgTypes.toSet) {
+          grpcSinkCalls += sinkCall
+        }
+      })
+
+    })
+
+    return grpcSinkCalls
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala:48]: @gaurav-gogia don't return ListBuffer, instead return List
add `.l` at the end of grpcSinkCalls

File: src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  def getGrpcSinkRules(cpg: Cpg): ListBuffer[RuleInfo] = {
+    val endpoints = getGrpcEndpoints(cpg)
+    var ruleList  = ListBuffer[RuleInfo]()
+
+    getGrpcSinkCalls(cpg, endpoints).foreach(sink => {
+      ruleList += RuleInfo(
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala:56]: @gaurav-gogia You are creating n rules with the same ruleId, which will not help.

use something like this 
`val methodFullNamesCombined = getGrpcSinkCalls(cpg, endpoints).map(sink => sink.methodFullName).mkString("(","|",")")`

and then add this to the RulePattern

File: src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          grpcSinkCalls += sinkCall
+        }
+      })
+
+    })
+
+    return grpcSinkCalls
+  }
+
+  def getGrpcSinkRules(cpg: Cpg): ListBuffer[RuleInfo] = {
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala:51]: Again return List

**Review by khemrajrathore — COMMENTED** [review]
@gaurav-gogia Have you tested your changes, are you getting desired results?

File: src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      // Detecting `onNext` call inside of gRPC endpoint method
+      // `onNext` takes on argument, this should be the call that satisfies gRPC/proto file contract
+      // Get full type name of arguments that the server/endpoint will process
+      val endpointArgTypes =
+        endpoint.call.methodFullName(OnNextPattern).argument.isCall.argument.isIdentifier.typeFullName.l
+
+      callList.foreach(sinkCall => {
+        // Get full type name of arguments going inside gRPC sink call
+        val sinkArgTypes = sinkCall.argument.isIdentifier.typeFullName.l
+        if (endpointArgTypes.toSet == sinkArgTypes.toSet) {
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala:41]: use `equals()` instead

**Comment by gaurav-gogia** [comment]
> Have you tested your changes, are you getting desired results?

@khemrajrathore yes, I've tested it on some sample repos. Have linked sample code in description

File: src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    return RuleInfo(
+      "Sinks.RPC.gRPC.Call",
+      "gRPC Call",
+      "",
+      Array[String]("grpc.io"),
+      List[String](methodFullNamesCombined),
+      false,
+      "",
+      HashMap[String, String](),
+      NodeType.REGULAR,
```
Comment (khemrajrathore) [inline src/main/scala/ai/privado/languageEngine/java/tagger/Utility/GRPCTaggerUtility.scala:84]: @gaurav-gogia Change this to NodeType.API

**Comment by gaurav-gogia** [comment]
> @gaurav-gogia Have you tested your changes, are you getting desired results?

yes, tested with [grpc-java](https://github.com/grpc/grpc-java) repo

**Pull request merged**